src/test/java/net/minecraftforge/debug/DimensionTest.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2020.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.debug;
+
+import java.util.function.Supplier;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.material.Material;
+import net.minecraft.data.DataGenerator;
+import net.minecraft.item.BlockItem;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemGroup;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.world.Dimension;
+import net.minecraft.world.biome.provider.OverworldBiomeProvider;
+import net.minecraft.world.gen.DimensionSettings;
+import net.minecraft.world.gen.NoiseChunkGenerator;
+import net.minecraft.world.gen.settings.DimensionGeneratorSettings;
+import net.minecraft.world.gen.settings.DimensionStructuresSettings;
+import net.minecraftforge.client.model.generators.BlockStateProvider;
+import net.minecraftforge.client.model.generators.ModelFile;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.dimension.BiomeProviders;
+import net.minecraftforge.dimension.DynamicDimensionManager;
+import net.minecraftforge.event.dimension.DimensionRegisterEvent;
+import net.minecraftforge.eventbus.api.IEventBus;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import net.minecraftforge.fml.RegistryObject;
+import net.minecraftforge.fml.common.Mod;
+import net.minecraftforge.fml.common.Mod.EventBusSubscriber.Bus;
+import net.minecraftforge.fml.event.lifecycle.GatherDataEvent;
+import net.minecraftforge.fml.event.server.FMLServerStartingEvent;
+import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
+import net.minecraftforge.registries.DeferredRegister;
+import net.minecraftforge.registries.ForgeRegistries;
+import net.minecraftforge.registries.ForgeRegistryEntry;
+import net.minecraftforge.registries.IForgeRegistry;
+import net.minecraftforge.registries.RegistryBuilder;
+
+@SuppressWarnings("unused")
+@Mod(DimensionTest.MODID)
+@Mod.EventBusSubscriber(bus = Bus.FORGE)
+public class DimensionTest {
```

Comment:
you only create a dimension in the event you post before the server is created and when the server is starting, i fail to grasp how this is a "dynamic" dimension creation. I would imagine this should be able to create a dimension/world at any point during the game. Also, if this system is capable of creating dimensions on the fly, I would suggest stress testing this, by seeing how much time it takes to create different dimensions (i.e. a second overworld vs a second End) or when creating 2 dimensions at once. I would assume that creating 2 dims at once is pretty heavy and would require some sort of work queue, where they arent processed all at the same time and dont cause the game to freeze forever.

Reply:
Thanks for your comment. I used the ServerStartingEvent because it is called exactly after spawn chunks are loaded and the Done message pops up in server log(at least on Dedicated). Therefore I thought it was suitable for testing it after world loading. Of course it should work outside of the ServerStartingEvent too, but I thought this was plausible with this example. The dimension adding shouldn't be too heavy on the server. The server will just tick it in the next tick, when creating a dimension, loading chunks as necessary when the player gets ported there or when a chunk is forced. I don't think that anyone will ever create more than 2 dimensions at once. That wouldn't make any sense.

Reply:
ah okay, since it is a one off thing per world load it did not seem quite dynamic.
and in a server it could very well happen, 2 players using RF Dimensions and creating one at the same time or similar times. I would just debug it to make sure. This is some pretty heavy stuff and has a lot of potential for problems as i understand it. The more you test the more seriously this can be considered. However this is just my opinion.

Reply:
From my tests, it takes about 6 seconds per dimension to create it. A reasonable amount of time for creating a load of chunks, though if you want, I can add that it forces the spawn chunks to be loaded at world creation. When both worlds are registered at the same tick they might get created both.